File: mana/lib/mana/repo_compass/formatter.ex
defmodule Mana.RepoCompass.Formatter do
  @moduledoc "Formats indexed project symbols for prompt injection"

  @max_chars 2400

  @doc "Format project index as a prompt section"
  @spec format([map()], String.t()) :: String.t()
  def format(index, project_name) do
    header = "## Repo Compass\nProject: #{project_name}\nStructural context map:\n"

    lines =
      Enum.map(index, fn %{path: path, kind: kind, symbols: symbols} ->
        symbol_str = Enum.join(symbols, "; ")
        "- #{path} [#{format_kind(kind)}]: #{symbol_str}"
      end)

    full = header <> Enum.join(lines, "\n")

    if String.length(full) > @max_chars do
      truncate(full, @max_chars)
    else
      full
    end
  end

  defp format_kind(:elixir_module), do: "elixir"
  defp format_kind(:elixir_script), do: "script"
  defp format_kind(:python_module), do: "python"
  defp format_kind(:documentation), do: "docs"
  defp format_kind(:config), do: "config"
  defp format_kind(_), do: "file"

  defp truncate(text, max) do
    String.slice(text, 0, max - 3) <> "..."
  end
end

File: code_puppy/plugins/repo_compass/formatter.py
from __future__ import annotations

from pathlib import Path

from .turbo_indexer_bridge import FileSummary


def _detect_project_name(root: Path) -> str:
    return root.name or "project"


def format_structure_map(
    root: Path, summaries: list[FileSummary], max_chars: int = 2400
) -> str | None:
    if not summaries:
        return None

    lines = [
        "## Repo Compass",
        f"Project: {_detect_project_name(root)}",
        "Structural context map:",
    ]

    for summary in summaries:
        bullet = f"- {summary.path} [{summary.kind}]"
        if summary.symbols:
            bullet += f": {'; '.join(summary.symbols)}"
        lines.append(bullet)

    text = "\n".join(lines)
    if len(text) <= max_chars:
        return text

    trimmed: list[str] = lines[:3]
    for line in lines[3:]:
        if len("\n".join(trimmed + [line, "- ... truncated ..."])) > max_chars:
            trimmed.append("- ... truncated ...")
            break
        trimmed.append(line)
    return "\n".join(trimmed)
